pkg/controller/labelidentity/label_group_index.go

```diff
@@ -0,0 +1,503 @@
+	// DeleteLabelIdentity deletes a LabelIdentity from the index.
+	DeleteLabelIdentity(labelKey string)
+	// AddEventHandler registers an eventHandler with the index.
+	AddEventHandler(handler eventHandler)
+	// Run starts the index.
+	Run(stopCh <-chan struct{})
+	// HasSynced returns true if the interface has been initialized with the full lists of LabelIdentities.
+	HasSynced() bool
+}
+
+type selectorItemUpdateEvent string
+
+const (
+	selectorMatchedLabelAdd     selectorItemUpdateEvent = "labelAdd"
+	selectorMatchedLabelDelete  selectorItemUpdateEvent = "labelDelete"
+	selectorMatchedPolicyAdd    selectorItemUpdateEvent = "policyAdd"
+	selectorMatchedPolicyDelete selectorItemUpdateEvent = "policyDelete"
+)
+
+// selectorItem represents a ClusterSet-scope selector from Antrea-native policies.
+// It also stores the LabelIdentity keys that this selector currently selects, as well
+// as the keys of Antrea-native policies that have this selector.
+type selectorItem struct {
+	selector *types.GroupSelector
+	// Keys are the normalized labels of matching LabelIdentities
+	labelIdentityKeys sets.String
+	// Keys are the UIDs of the policies that have the selector in their specs.
+	policyKeys sets.String
+}
+
+func (s *selectorItem) getKey() string {
+	return s.selector.NormalizedName
+}
+
+// labelIdentityMatch is constructed from a LabelIdentity and used for matching
+// between LabelIdentity and selectorItems. It also stores the current selectorItems
+// that matches this LabelIdentity.
+type labelIdentityMatch struct {
+	id               uint32
+	namespace        string
+	namespaceLabels  map[string]string
+	podLabels        map[string]string
+	selectorItemKeys sets.String
+}
+
+// matches knows if a LabelIdentity matches a selectorItem.
+func (l *labelIdentityMatch) matches(s *selectorItem) bool {
+	selectorItemNamespace := s.selector.Namespace
+	if selectorItemNamespace != emptyNamespace && selectorItemNamespace != l.namespace {
+		return false
+	}
+	if s.selector.NamespaceSelector != nil && !s.selector.NamespaceSelector.Matches(labels.Set(l.namespaceLabels)) {
+		return false
+	}
+	// At this stage Namespace has matched
+	if s.selector.PodSelector != nil {
+		return s.selector.PodSelector.Matches(labels.Set(l.podLabels))
+	}
+	// SelectorItem selects all when all selectors are missing.
+	return true
+}
+
+// constructMapFromLabelString parses label string of format "app=client,env=dev" into a map.
+func constructMapFromLabelString(s string) map[string]string {
+	m := map[string]string{}
+	kvs := strings.Split(s, ",")
+	for _, kv := range kvs {
+		kvpair := strings.Split(kv, "=")
+		m[kvpair[0]] = kvpair[1]
+	}
+	return m
+}
+
+// newLabelIdentityMatch constructs a labelIdentityMatch from a normalized LabelIdentity string.
+func newLabelIdentityMatch(labelIdentity string, id uint32) *labelIdentityMatch {
+	nsIndex, podIndex := labelRegex.SubexpIndex("nslabels"), labelRegex.SubexpIndex("podlabels")
```

Comment:
Same comment as https://github.com/antrea-io/antrea/pull/3913#discussion_r993580635, the repeated overhead could be avoided by using global variables since they are not specific to arguments.

Reply:
Done